Add Hierarchical Stats example

apatole opened the pull request:
### Description

This patch adds hierarchical stats example and the related notebooks to try it out.

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [x] Quick tests passed locally by running `./runtest.sh`.
- [x] In-line docstrings updated.
- [x] Documentation updated.

apatole commented:
Hi @chesterxgchen @yanchengnv Please help review PR to add hierarchical statistics example, thanks.

apatole commented on examples/advanced/federated-statistics/README.md:
```diff
@@ -89,6 +89,40 @@ via the MONAI FLARE integration.
 load and cache the externally calculated statistics. The server side controller will pass the target metrics to pre_run method 
 so it can be used to load the statistics. 
 
+### Hierarchical Stats example
```
Done, thanks.

apatole commented:
Thanks @chesterxgchen for the Review!
I have incorporated review comments and uploaded a new patch set.

YuanTingHsieh commented on examples/advanced/federated-statistics/hierarchical_stats.ipynb:
```diff
@@ -0,0 +1,236 @@
+{
+ "cells": [
```
suggest moving this notebook file inside the "hierarchical_stats" folder.

apatole replied:
Done, thanks.

YuanTingHsieh commented on examples/advanced/federated-statistics/README.md:
```diff
@@ -89,6 +89,72 @@ via the MONAI FLARE integration.
+
+Expanding `Global` will show visualization of global statistics similar to the following
+
+<img src="hierarchical_stats/demo/global_tats.png" alt="Global stats" width="800">
```
typo on file name global_tats.png -> global_stats.png

YuanTingHsieh commented on examples/advanced/federated-statistics/hierarchical_stats/README.md:
```diff
@@ -0,0 +1,245 @@
+
+Expanding `Global` will show visualization of global statistics similar to the following
+
+<img src="demo/global_tats.png" alt="Global stats" width="400">
```
typo on file name global_tats.png -> global_stats.png

apatole replied:
Done, thanks.

YuanTingHsieh commented on examples/advanced/federated-statistics/hierarchical_stats/demo/global_tats.png:
typo on file name global_tats.png -> global_stats.png

YuanTingHsieh commented on examples/advanced/federated-statistics/hierarchical_stats.ipynb:
```diff
@@ -0,0 +1,236 @@
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "We are not quite done yet. What if you prefer to use python API instead CLI to run jobs. Lets do that in this section"
+   ]
```
this line confuses me a bit, what "section" are we referring to here?

apatole replied:
Actually that line got copied from the other notebook I referred. Removed it now as I don't see it relevant here. Thanks.

YuanTingHsieh commented on examples/advanced/federated-statistics/hierarchical_stats.ipynb:
```diff
@@ -0,0 +1,236 @@
+   },
+   "source": [
+    "## Visualization\n",
+    "We can visualize the results easly via the visualizaiton notebook. Before we do that, we need to copy the data to the notebook directory \n"
```
typo: visualizaiton -> visualization

apatole replied:
Done, thanks.

YuanTingHsieh commented on examples/advanced/federated-statistics/hierarchical_stats/README.md:
```diff
@@ -0,0 +1,245 @@
+        "writer_id": "stats_writer",
+        "enable_pre_run_task": false,
+        "precision": 4,
+        "hierarchy_config": "hierarchy_config.json"
```
could you add some explanation on any requirements/structure of this "hierarchy_config"?
Does it have to be a JSON file? Is there any key that must exist etc?

apatole replied:
Done, thanks. Sorry I missed this important part earlier.

YuanTingHsieh commented on examples/advanced/federated-statistics/hierarchical_stats/demo/visualization.ipynb:
```diff
@@ -0,0 +1,222 @@
+   "source": [
+    "#### dependencies\n",
+    "\n",
+    "To run examples, you might need to install the dependencies\n",
+    "* ipywidgets\n",
+    "* numpy\n",
+    "* pandas\n",
+    "* matplotlib\n",
+    "* jupyter\n",
+    "\n",
+    "These are captured in the requirements.txt"
```
can we do pip install ../requirements.txt as you did in the other notebook?

apatole replied:
Done, thanks.

YuanTingHsieh commented on examples/advanced/federated-statistics/hierarchical_stats/demo/visualization.ipynb:
```diff
@@ -0,0 +1,222 @@
+   "source": [
+    "### Overall Statistics\n",
+    "\n",
+    "In this section, we demonstrate how to visualize hierarchical stats from the tabular data."
```
suggest: "In this section, we demonstrate how to visualize hierarchical stats from the tabular data using Jupyter Widgets."

apatole replied:
Done, thanks.

YuanTingHsieh commented on examples/advanced/federated-statistics/hierarchical_stats/demo/visualization.ipynb:
```diff
@@ -0,0 +1,222 @@
+   "id": "3e86860e",
+   "metadata": {},
+   "source": [
+    "The following code demonstrates how the hierarchical histograms can be displayed display for for features like `percentage`."
```
suggest "The following code demonstrates how the hierarchical histograms can be displayed using Jupyter Widgets for features like `percentage`."

apatole replied:
Done, thanks.

apatole commented on examples/advanced/federated-statistics/hierarchical_stats.ipynb:
```diff
@@ -0,0 +1,236 @@
+   "source": [
+    "# Federated Hierarchical Statistics \n",
+    "\n",
+    "In this example, we will show how to generate federated Hierarchical statistics for data that can be represented as Pandas Data Frame.\n",
```
Done, thanks.

apatole commented on examples/advanced/federated-statistics/hierarchical_stats/utils/prepare_data.py:
```diff
@@ -0,0 +1,70 @@
+import csv
+import os
+import random
+import sys
```
Done, thanks.

apatole commented:
Many thanks @yanchengnv for the review, I have uploaded updated patch incorporating review comments.

Couple of  additional changes:

- Updated `examples/advanced/federated-statistics/README.md` with the link to hierarchical stats.
- Fixed existing broken link in examples/advanced/federated-statistics/README.md

YuanTingHsieh reviewed (approved):
thanks, LGTM

SYangster commented:
/build

apatole commented:
Thanks @SYangster for triggering a build and helping with `blossom-ci`, all checks now passed.

Looks like someone with the write access need to merge it.
